### PR TITLE
[WIP] [Issue 131] - Implement kernel connection file reset if fails to start for the first time

### LINKED ISSUE #131: Kernel launch failure when address is in use
Rarely when running the jupyter/kernel_gateway test suite, the tests fail with the following exception due to a race condition between two kernels using the same port:

```
Traceback (most recent call last):
  File "/opt/python/3.3.5/lib/python3.3/runpy.py", line 160, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/opt/python/3.3.5/lib/python3.3/runpy.py", line 73, in _run_code
    exec(code, run_globals)
  File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/ipykernel/__main__.py", line 3, in <module>
    app.launch_new_instance()
  File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/traitlets/config/application.py", line 595, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-122>", line 2, in initialize
  File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/traitlets/config/application.py", line 74, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/ipykernel/kernelapp.py", line 412, in initialize
    self.init_sockets()
  File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/ipykernel/kernelapp.py", line 245, in init_sockets
    self.init_iopub(context)
  File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/ipykernel/kernelapp.py", line 250, in init_iopub
    self.iopub_port = self._bind_socket(self.iopub_socket, self.iopub_port)
  File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/ipykernel/kernelapp.py", line 174, in _bind_socket
    s.bind("tcp://%s:%i" % (self.ip, port))
  File "zmq/backend/cython/socket.pyx", line 487, in zmq.backend.cython.socket.Socket.bind (zmq/backend/cython/socket.c:5156)
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:7535)
zmq.error.ZMQError: Address already in use
```

IIRC, the port assignment is requested by a client (e.g., via a connection file). Having the kernel retry on another port means the client won't be able to communicate with that kernel.

Is this behavior spec'ed somewhere? Is it the client's responsibility to retry (e.g., this is why I see the rare "kernel has died, restarting" flow sometimes when first opening a notebook?) Can the behavior be improved?

/cc @minrk

### PR DESCRIPTION
**Pull request by poplav**
Fixes #131

### CODE REVIEW

**Comment by parente** [comment]
Please don't merge yet. Travis port conflicts are the most reliable way for us to test this fix.

**Comment by parente** [comment]
No go: https://api.travis-ci.org/jobs/117977967/log.txt?deansi=true

```
The command "jupyter kernelgateway --help" exited with 0.
70.12s$ nosetests -x --process-restartworker --with-coverage --cover-package=kernel_gateway
...............................................................[KernelGatewayApp] Kernel shutdown: 84b3a547-0b10-42d3-a983-c509d672f9b2
[KernelGatewayApp] Kernel shutdown: 4b79d566-aa2d-48e0-93f4-bc43ac097f41
...............................Traceback (most recent call last):
  File "/opt/python/3.4.2/lib/python3.4/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/python/3.4.2/lib/python3.4/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/ipykernel/__main__.py", line 3, in <module>
    app.launch_new_instance()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/traitlets/config/application.py", line 595, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-122>", line 2, in initialize
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/traitlets/config/application.py", line 74, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/ipykernel/kernelapp.py", line 412, in initialize
    self.init_sockets()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/ipykernel/kernelapp.py", line 232, in init_sockets
    self.shell_port = self._bind_socket(self.shell_socket, self.shell_port)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/ipykernel/kernelapp.py", line 174, in _bind_socket
    s.bind("tcp://%s:%i" % (self.ip, port))
  File "zmq/backend/cython/socket.pyx", line 487, in zmq.backend.cython.socket.Socket.bind (zmq/backend/cython/socket.c:5156)
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:7535)
zmq.error.ZMQError: Address already in use
E
======================================================================
ERROR: Notebook source should exist under the path `/_api/source`.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/tornado/testing.py", line 376, in setUp
    self._app = self.get_app()
  File "/home/travis/build/jupyter/kernel_gateway/kernel_gateway/tests/test_gatewayapp.py", line 90, in get_app
    self.app.init_configurables()
  File "/home/travis/build/jupyter/kernel_gateway/kernel_gateway/gatewayapp.py", line 289, in init_configurables
    self.kernel_manager
  File "/home/travis/build/jupyter/kernel_gateway/kernel_gateway/services/kernels/pool.py", line 65, in __init__
    super(ManagedKernelPool, self).__init__(prespawn_count, kernel_manager)
  File "/home/travis/build/jupyter/kernel_gateway/kernel_gateway/services/kernels/pool.py", line 27, in __init__
    self.kernel_manager.start_seeded_kernel()
  File "/home/travis/build/jupyter/kernel_gateway/kernel_gateway/services/kernels/manager.py", line 63, in start_seeded_kernel
    self.start_kernel(kernel_name=self.seed_kernelspec, *args, **kwargs)
  File "/home/travis/build/jupyter/kernel_gateway/kernel_gateway/services/kernels/manager.py", line 79, in start_kernel
    client.wait_for_ready()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/jupyter_client/blocking/client.py", line 59, in wait_for_ready
    raise RuntimeError('Kernel died before replying to kernel_info')
nose.proxy.RuntimeError: Kernel died before replying to kernel_info
-------------------- >> begin captured logging << --------------------
root: DEBUG: Native kernel (python3) available from /home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/ipykernel/resources
root: DEBUG: Native kernel (python3) available from /home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/ipykernel/resources
root: DEBUG: connecting shell channel to tcp://127.0.0.1:42454
root: DEBUG: Connecting to: tcp://127.0.0.1:42454
root: DEBUG: connecting iopub channel to tcp://127.0.0.1:47746
root: DEBUG: Connecting to: tcp://127.0.0.1:47746
root: DEBUG: connecting stdin channel to tcp://127.0.0.1:49857
root: DEBUG: Connecting to: tcp://127.0.0.1:49857
root: DEBUG: connecting heartbeat channel to tcp://127.0.0.1:60545
--------------------- >> end captured logging << ---------------------
```

The kernel bombs with an address in use error. jupyter_client raises a runtime error as a result. The `restart_kernel` method never even gets invoked at this point since an exception bubbles out of the initial launch.

I think the logic for handling initial, full restarts needs to be in `start_kernel`, not in `restart_kernel`.

**Comment by poplav** [comment]
Moved the restart logic to start_kernel, where it will shutdown and recursively call start again if the runtime error is raised as in the error above.  I still saw an [error] (https://s3.amazonaws.com/archive.travis-ci.org/jobs/118258079/log.txt) yesterday, but that was the only instance after ~50 attempts.  The issue is difficult to reproduce.

File: kernel_gateway/services/kernels/manager.py [hunk 1]
```diff
@@ -76,7 +76,16 @@ def start_kernel(self, *args, **kwargs):
                 client = kernel.client()
                 # Only start channels and wait for ready in HTTP mode
                 client.start_channels()
-                client.wait_for_ready()
+                try:
+                    client.wait_for_ready()
```
Comment (parente) [inline kernel_gateway/services/kernels/manager.py]: This code only runs when there's a `seed_source` notebook configured. What if there isn't (which is the far more typical case)? The parent class `start_kernel` call is what actually spawns the kernel. Is kernel failure with the address error synchronous with that call and we can catch it there? Or not, and then we've got the problem of the client dying asynchronously at any point in the code base unless we wait for it before returning from `start_kernel` which is a new behavior.

Comment (poplav) [inline kernel_gateway/services/kernels/manager.py]: With [wait_for_ready](https://github.com/jupyter/jupyter_client/blob/637fb55afe9989b4a5f420d5e32a5a11e907f770/jupyter_client/blocking/client.py#L23) it is synchronous.  We could then init the client and check that it has started/is ready regardless of whether it has a seed_source (missed that part).  I would have thought the approach in the first commit would work though, as I think its better not to deal with the problem on this level.

**Comment by parente** [comment]
We haven't made any progress on fixing the kernel port conflict problem in a while. It hasn't been a huge issue during tests like it was originally for some reason. (Travis nodes are less busy?)

I'm going to close this PR as stale. If we decide to pick up the problem in the future, we can resume the hunt from here if it makes sense.

**Pull request closed without merging**